LFortran: add new package

certik opened the pull request:
For now only a Debug build. It installs the LFortran runtime library and
LFortran can compile codes to binaries. The interactive mode does not
work yet with LLVM > 11, that has to be fixed upstream.

certik commented:
This is ready for review.

awvwgk commented on var/spack/repos/builtin/packages/lfortran/package.py:
```diff
@@ -0,0 +1,50 @@
+from spack.package import *
+
+
+class Lfortran(Package):
```
Why not make this a `CMakePackage`?

awvwgk commented on var/spack/repos/builtin/packages/lfortran/package.py:
```diff
@@ -0,0 +1,50 @@
+    maintainers = ['certik']
+
+    def install(self, spec, prefix):
+        open("version", "w").write(str(self.version))
+
+        python = which("python")
+        python("grammar/asdl_cpp.py")
+        python("grammar/asdl_cpp.py", "src/libasr/ASR.asdl", "src/libasr/asr.h")
+
+        re2c = which("re2c")
+        bison = which("bison")
+        with working_dir("src/lfortran/parser"):
+            re2c("-W", "-b", "tokenizer.re", "-o", "tokenizer.cpp")
+            re2c("-W", "-b", "preprocessor.re", "-o", "preprocessor.cpp")
+            bison("-Wall", "-d", "-r", "all", "parser.yy")
```
Run those steps in the preparing / patching stage.

junghans replied:
Done!

awvwgk commented on var/spack/repos/builtin/packages/lfortran/package.py:
```diff
@@ -0,0 +1,50 @@
+
+
+class Lfortran(Package):
+    """LFortran"""
```
Could be a little more precise
```suggestion
    """Modern interactive LLVM-based Fortran compiler"""
```

awvwgk commented:
You can make the pre-build step version dependent in case this is necessary. Having the pre-build step would allow you to compile from the latest git source using spack, which is probably more convenient than only having releases available.

certik commented:
Yes, that was my motivation to allow building directly from git. Yes, I'll work on this. I'll try to rework upstream LFortran to have a robust multiplatform script that does the pre-build step, so we can just call it.

awvwgk commented:
@certik, feel free to reuse whatever is useful from my initial attempt in https://github.com/spack/spack/pull/28165

alalazo commented:
@certik Can you address review and make this a `CMakePackage`?

certik commented:
@alalazo I don't have the time right now. We are laser focused on delivering LFortran itself. This will have to wait a few more months, or others can help as well.

alalazo commented:
@certik If you're fine with it, and it helps, I can try to push here the modifications myself. If you prefer to wait and finish the PR that would be fine too.

certik commented:
@alalazo go ahead and push them here. Thank you! I don't have time right now to work on it. If you discover anything that we need to fix in our build system, let me know.

certik commented:
Thanks @junghans !

junghans commented:
Started on the CMakePackage migration

junghans commented on var/spack/repos/builtin/packages/lfortran/package.py:
```diff
@@ -0,0 +1,47 @@
+    maintainers = ["certik"]
+
+    @run_before("cmake")
+    def prepare_source(self):
```
@adamjstewart any ideas how to write this nicer?

adamjstewart replied:
Seems fine to me. You could also create a new phase or put it in patch but I think I like this better

becker33 commented on var/spack/repos/builtin/packages/lfortran/package.py:
```diff
@@ -0,0 +1,47 @@
+    def prepare_source(self):
+        open("version", "w").write(str(self.version))
+
+        python = which("python")
```
```suggestion
        python = self.spec["python'].command
```
This will eliminate any issue of `python` vs `python3`

becker33 commented on var/spack/repos/builtin/packages/lfortran/package.py:
```diff
@@ -0,0 +1,47 @@
+            bison("-Wall", "-d", "-r", "all", "parser.yy")
+
+    def cmake_args(self):
+        args = ["-DWITH_LLVM=yes", "-DLFORTRAN_BUILD_ALL=yes", "-DWITH_STACKTRACE=yes"]
```
These look like future candidates for variants?

certik replied:
Yes, the LLVM and STACKTRACE can be on/off and there are some other things that can be turned on and off.

becker33 commented on var/spack/repos/builtin/packages/lfortran/package.py:
```diff
@@ -0,0 +1,47 @@
+    git = "https://gitlab.com/lfortran/lfortran.git"
+
+    version("master", branch="master")
+    version("0.16.0", tag="v0.16.0")
+    version("0.15.0", tag="v0.15.0")
```
Would using the release tarballs for these two versions mean we only have to do the fancy stuff below for GitVersion versions?

certik replied:
Yes, correct. The tarballs have everything generated, so only depend on cmake, not Bison, re2c or Python.

junghans replied:
Where are the tarball?

junghans replied:
@certik where are the tarballs?

certik replied:
Ah, here is the link to all of them: https://lfortran.org/download/

junghans replied:
Ok let me work that in....

junghans replied:
Done!

junghans commented:
@certik I get the following build error:
```
[ 15%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/global_stmts_program.cpp.o
cd /tmp/junghans/spack-stage/spack-stage-lfortran-0.16.0-o2aqz4fuopw2ur7ybv7hvgo5pos4ur4c/spack-build-o2aqz4f/src/libasr && /vast/home/junghans/spack.git/lib/spack/env/gcc/g++ -DHAVE_TARGET_X86=1 -DHAVE_WHEREAMI=1 -DHAVE_ZLIB=1 -I/tmp/junghans/spack-stage/spack-stage-lfortran-0.16.0-o2aqz4fuopw2ur7ybv7hvgo5pos4ur4c/spack-build-o2aqz4f/src/libasr/.. -I/tmp/junghans/spack-stage/spack-stage-lfortran-0.16.0-o2aqz4fuopw2ur7ybv7hvgo5pos4ur4c/spack-src/src/libasr/.. -isystem /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/include -isystem /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/include -isystem /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/zlib-1.2.12-bea67asvdme72ijc7c43f47ybecnhvkh/include -Wall -Wextra -O3 -march=native -funroll-loops -DNDEBUG -std=gnu++17 -fPIC -fno-rtti -MD -MT src/libasr/CMakeFiles/asr.dir/pass/global_stmts_program.cpp.o -MF CMakeFiles/asr.dir/pass/global_stmts_program.cpp.o.d -o CMakeFiles/asr.dir/pass/global_stmts_program.cpp.o -c /tmp/junghans/spack-stage/spack-stage-lfortran-0.16.0-o2aqz4fuopw2ur7ybv7hvgo5pos4ur4c/spack-src/src/libasr/pass/global_stmts_program.cpp
In file included from /tmp/junghans/spack-stage/spack-stage-lfortran-0.16.0-o2aqz4fuopw2ur7ybv7hvgo5pos4ur4c/spack-src/src/libasr/codegen/wasm_to_wat.cpp:4:
/tmp/junghans/spack-stage/spack-stage-lfortran-0.16.0-o2aqz4fuopw2ur7ybv7hvgo5pos4ur4c/spack-src/src/libasr/../libasr/codegen/wasm_to_wat.h:4:10: fatal error: libasr/wasm_visitor.h: No such file or directory
 #include <libasr/wasm_visitor.h>
          ^~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [src/libasr/CMakeFiles/asr.dir/build.make:163: src/libasr/CMakeFiles/asr.dir/codegen/wasm_to_wat.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

certik commented:
The `wasm_visitor.h` is generated by our `build0.sh` script: https://github.com/lfortran/lfortran/blob/bb0307908559e79904e794e71a452d026a176e6a/build0.sh#L14. The Spack build system needs to execute all parts of the `build0.sh`. I think you just need to add this one line that I linked to.

certik commented on var/spack/repos/builtin/packages/lfortran/package.py:
```diff
@@ -0,0 +1,47 @@
+        python = self.spec["python'].command
+        python("grammar/asdl_cpp.py")
+        python("grammar/asdl_cpp.py", "src/libasr/ASR.asdl", "src/libasr/asr.h")
+
```
This should fix it:
```suggestion
        python("src/libasr/wasm_instructions_visitor.py")

```

junghans commented:
Got a bit further:
```
[ 94%] Linking CXX executable test_cwrapper
cd /tmp/junghans/spack-stage/spack-stage-lfortran-0.19.0-6spdt7undm5cbxfxxqkugddrw67iufiu/spack-build-6spdt7u/src/lfortran/tests && /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/cmake-3.26.3-p7ykx4mwtpzji7l4tk3ezhirgro
47yta/bin/cmake -E cmake_link_script CMakeFiles/test_cwrapper.dir/link.txt --verbose=1
/vast/home/junghans/spack.git/lib/spack/env/gcc/g++ -Wall -Wextra -O3 -funroll-loops -DNDEBUG CMakeFiles/test_cwrapper.dir/test_cwrapper.c.o -o test_cwrapper  ../liblfortran_lib.a ../../libasr/libasr.a ../../runtime/legacy/liblfortran_runtime
_static.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/lib/libbfd.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj
3ic2k5jywlca/lib/libLLVMMCJIT.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMOrcJIT.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-
14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMExecutionEngine.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMRuntimeDyld.a /vast/home/junghans/spack.git/opt
/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMPasses.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMCoroutines.a /
vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMipo.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jyw
lca/lib/libLLVMFrontendOpenMP.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMIRReader.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llv
m-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMLinker.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMObjCARCOpts.a /vast/home/junghans/spack.git/opt/spack/
linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMVectorize.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMJITLink.a /vast/ho
me/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMOrcTargetProcess.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic
2k5jywlca/lib/libLLVMOrcShared.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMAsmParser.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/l
lvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMX86CodeGen.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMAsmPrinter.a /vast/home/junghans/spack.git/opt/s
pack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMDebugInfoMSF.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMInstrument
ation.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMGlobalISel.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cd
rosqtxj3ic2k5jywlca/lib/libLLVMSelectionDAG.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMCodeGen.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/
gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMTarget.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMBitWriter.a /vast/home/junghans/spack.git
/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMScalarOpts.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMAggres
siveInstCombine.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMInstCombine.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4e
yc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMTransformUtils.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMAnalysis.a /vast/home/junghans/spack.git/opt/spack/linux-
rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMProfileData.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMDebugInfoDWARF.a /vast
/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMObject.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywl
ca/lib/libLLVMBitReader.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMTextAPI.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.
4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMCFGuard.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMCore.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-h
aswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMRemarks.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMBitstreamReader.a /vast/home/jung
hans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMX86AsmParser.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/l
ib/libLLVMX86Desc.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMMCParser.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4ey
c4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMX86Disassembler.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMX86Info.a /vast/home/junghans/spack.git/opt/spack/linux-r
hel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMMCDisassembler.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMMC.a /vast/home/jung
hans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMBinaryFormat.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/l
ib/libLLVMDebugInfoCodeView.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMSupport.a /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/zlib-1
.2.12-bea67asvdme72ijc7c43f47ybecnhvkh/lib/libz.a -lrt -ldl -lpthread -lm /usr/lib64/libtinfo.so /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/llvm-14.0.4-4eyc4qu2am6cdrosqtxj3ic2k5jywlca/lib/libLLVMDemangle.a
/vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/bin/ld: ../../libasr/libasr.a(asr_utils.cpp.o): in function `LCompilers::ASRUtils::find_and_load_module(Allocator&, std::__cx
x11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::SymbolTable&, bool, LCompilers::PassOptions&)':
asr_utils.cpp:(.text+0x1c3c): undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
/vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/bin/ld: asr_utils.cpp:(.text+0x1d7c): undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
/vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/bin/ld: asr_utils.cpp:(.text+0x2797): undefined reference to `std::filesystem::__cxx11::path::has_root_directory() const'
/vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/bin/ld: asr_utils.cpp:(.text+0x2b4d): undefined reference to `std::filesystem::__cxx11::path::has_filename() const'
/vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/bin/ld: asr_utils.cpp:(.text+0x2be4): undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
/vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/bin/ld: asr_utils.cpp:(.text+0x30d3): undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
/vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/bin/ld: asr_utils.cpp:(.text+0x332a): undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
/vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/bin/ld: asr_utils.cpp:(.text+0x346c): undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
/vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/bin/ld: asr_utils.cpp:(.text+0x3484): undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
/vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/bin/ld: ../../libasr/libasr.a(asr_utils.cpp.o):asr_utils.cpp:(.text+0x34ac): more undefined references to `std::filesystem::_
_cxx11::path::_M_split_cmpts()' follow
/vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/bin/ld: /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/
lib/libbfd.a(tekhex.o): in function `tekhex_write_object_contents':
tekhex.c:(.text+0x5c3): undefined reference to `hex_init'
/vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/bin/ld: /vast/home/junghans/spack.git/opt/spack/linux-rhel8-haswell/gcc-8.5.0/binutils-2.38-o3yadkqfcroullmegfj75oqd27b5vtch/
lib/libbfd.a(tekhex.o): in function `first_phase':
tekhex.c:(.text+0x13e8): undefined reference to `_hex_value'
```

certik commented:
It's the `std::filesystem` again. People keep convincing me that it is now well supported on all platforms, but then I always get burned by it. What C++ compiler did you use and what version?

junghans commented:
linux-rhel8-haswell/gcc-8.5.0

junghans commented:
Look a lot like https://github.com/google/honggfuzz/issues/55 hinting an too old binutils....

certik committed:
Add LFortran

It installs the LFortran runtime library and
LFortran can compile codes to binaries. The interactive mode does not
work yet with LLVM > 11, that has to be fixed upstream.

Co-authored-by: Wileam Y. Phan <50928756+wyphan@users.noreply.github.com>

junghans commented:
I tried gcc-9.4.0, but same error, it seems to me the glibc on the system might be too old, which is something that spack cannot fix.

junghans commented on var/spack/repos/builtin/packages/lfortran/package.py:
```diff
@@ -0,0 +1,44 @@
+
+    @run_before("cmake")
+    def prepare_source(self):
+        if self.spec.satisfies("@main"):
```
@adamjstewart @becker33 is there a way to have two @ attributes? i.e. `@run_before("cmake")` and `@when("@main")` ?

adamjstewart replied:
@alalazo would be the expert here. I would try it and try both orders. It is sensitive to order.

junghans replied:
I did, but it didn't work! @alalazo any ideas?

junghans replied:
Ok, I just learned that `LFORTRAN_BUILD_ALL=no` runs `build0.sh` as part of CMake, so we can simplify this even more! @certik chime in if I am wrong here.

certik replied:
I think this is correct.

junghans commented:
> I tried gcc-9.4.0, but same error, it seems to me the glibc on the system might be too old, which is something that spack cannot fix.

I can confirm on a system with a newer glibc it just works! @becker33 @adamjstewart is there a ways to do something like `conflicts(glibc@:2.36)` ?

adamjstewart commented:
Not at the moment

junghans commented:
Ok, then this is ready for the final review!

junghans commented:
@certik I made the dependencies a bit less strict, please check if I got it right!

certik commented on var/spack/repos/builtin/packages/lfortran/package.py:
```diff
@@ -23,11 +23,11 @@ class Lfortran(CMakePackage):
+    depends_on("llvm@11:15", type="build", when="+llvm")
+    depends_on("zlib", type="build")
+    depends_on("re2c", type="build", when="@main")
+    depends_on("bison", type="build", when="@main")
```
You have to use Bison 3.4, the newer Bison packages have a bug that fails to compile our grammar correctly.

junghans replied:
Got it!

certik commented:
Otherwise the new versions of the dependencies should be fine.

junghans commented:
@awvwgk @adamjstewart @becker33 can you review this?

adamjstewart commented:
I don't know much about CMake, I'll let someone else review.

awvwgk reviewed (commented):
Now the package looks much cleaner, hope that LFortran can stay a well-behaved CMake project. Looks good to me.

certik commented:
> hope that LFortran can stay a well-behaved CMake project.

Heh, we are trying!

wyphan reviewed (approved):
Just successfully Spack-built LFortran 0.19 on my dev rig!

wyphan commented:
Welp, I guess I spoke too soon. While `lfortran --version` works just fine, compiling a "Hello, world!" Fortran code fails with LFortran being unable to find `clang`. I'll re-review and add the needed adjustments as a comment

wyphan commented on var/spack/repos/builtin/packages/lfortran/package.py:
```diff
@@ -0,0 +1,41 @@
+
+    depends_on("python@3:", type="build", when="@main")
+    depends_on("cmake", type="build")
+    depends_on("llvm@11:15", type="build", when="+llvm")
```
Since LFortran depends on `clang` to compile Fortran code, the "run" dependency type should be added for LLVM.
```suggestion
    depends_on("llvm@11:15", type=("build", "run"), when="+llvm")
```

wyphan reviewed (changes requested):
Try adjusting the Spack recipe as suggested below.

certik commented:
Yes, currently you need either clang or gcc to link. We'll later use the LLVM linker itself, that should remove all dependencies on clang/gcc, but for now you have to have them. However you can use any gcc or clang, it doesn't have to be the one used to build LFortran. You just need a linker.

wyphan reviewed (approved):
Confirmed working for both 0.19.0 and main branch on my dev rig.

haampie commented on var/spack/repos/builtin/packages/lfortran/package.py:
```diff
@@ -0,0 +1,41 @@
+    depends_on("python@3:", type="build", when="@main")
+    depends_on("cmake", type="build")
+    depends_on("llvm@11:15", type=("build", "run"), when="+llvm")
+    depends_on("zlib", type="build")
```
This is probably not correct? type=build?

junghans replied:
It is, because it is linked in statically only.

haampie reviewed (commented):
ping @certik
